sys/sched.c

```diff
@@ -159,6 +159,9 @@ long sched_switch(void) {
   /* make sure we reacquire td_spin lock on return to current context */
   td->td_flags |= TDF_NEEDLOCK;
 
+  if (PCPU_GET(no_switch))
+    panic("Switching context while interrupts are disabled is forbidden!");
+
```

Comment:
IMO this will never fail. Currently we set `no_switch` to true only when interrupts are disabled. And when interrupts are disabled, and `mips_exc_handler` will be called again (due to exception), there is no chance that we will context-switch as there is no `on_exc_leave` call in case 2b & 2c.

Reply:
Please note that someone, by accident, may attempt to use sleep mutex in:
1) Interrupt filter code
2) Exception handler code (esp. `tlb_exception_handler`)

We don't know if it's really the case, but it's better to be safe than sorry. AFAIR I've seen similar check in _FreeBSD_ kernel.

Reply:
Yes, alright, I haven't thought about calling `sched_switch` from turnstile or sleepqueue.

Reply:
Should I add some comment here to make the case more clear?

Reply:
I don't think so. The code is clear. It's just my mind that was fixed on checking interlacing of `mips_exc_handler` calls. :wink: